### PR TITLE
Add products source to Volur SDK

### PR DESCRIPTION
**Pull request by rebwibsey**
Add Product Source and tests to Völur SDK

Closes https://github.com/volur-ai/data-engineering/issues/2222

### CODE REVIEW

File: src/volur/sdk/v1alpha2/sources/csv/source.py [hunk 1]
```diff
@@ -250,3 +252,206 @@ def _create_material(
...
+                    "description",
+                    "string",
+                    characteristic_name="description",
+                ),
+                CharacteristicColumn(
+                    "plant_number",
+                    "integer",
+                    characteristic_name="plant_number",
+                ),
+            ],
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/sources/csv/source.py:301]: I think after this change we are going to use typed `CharacteristicColumns` https://github.com/volur-ai/python-volur-ai-sdk/blob/c42b6589ee1be71487502124ee08fba72d9a9766/src/volur/sdk/v1alpha2/sources/csv/base.py#L98-L251

Could you please update the example?

File: src/volur/api/v1alpha1/client.py [hunk 2]
```diff
@@ -111,3 +112,93 @@ async def generate_requests() -> (
...
+    ) -> Status:
+        """Uploads Products Information to the Völur platform using the Völur
+        API.
+
+        This method is using a source to get the products data and then send
+        it to the Völur API. This method is asynchronous and will return a
+        status of the operation.
+
+        Args:
+            products: a source of materials data to be uploaded to the Völur
```
Comment (nikitabarskov) [inline src/volur/api/v1alpha1/client.py:128]: Feel free to resolve

I would be picky and be careful with usage of a word `source`, would rather rephrase it this way:

```suggestion
            products: an iterable of Product protos to be uploaded via API
```

File: src/volur/sdk/v1alpha2/__init__.py [hunk 3]
```diff
@@ -1,5 +1,11 @@
 from volur.sdk.v1alpha2.client import VolurClient
+from volur.sdk.v1alpha2.sources import (
+    ProductsCSVFileSource,
+    ProductsSource,
+)
 
 __all__ = [
     "VolurClient",
+    "ProductsSource",
+    "ProductsCSVFileSource",
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/__init__.py:10]: Not required, we can keep them in `volur.sdk.v1alpha2.sources.csv`, if there is a need we can refactor later.

File: src/volur/sdk/v1alpha2/sources/__init__.py [hunk 4]
```diff
@@ -1 +1,9 @@
+from volur.sdk.v1alpha2.sources.csv import (
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/sources/__init__.py:1]:
```suggestion
from .csv import (
```

Relative imports are fine in this scenario, and it's less code.

File: src/volur/sdk/v1alpha2/sources/csv/__init__.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
+from volur.sdk.v1alpha2.sources.csv.base import ProductsSource
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/sources/csv/__init__.py:1]:
```suggestion
from .base import ProductsSource
```

Same as above

File: src/volur/sdk/v1alpha2/client.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 from loguru import logger
 from volur.api.v1alpha1.client import VolurApiAsyncClient
 from volur.sdk.v1alpha1.sources.csv.base import MaterialsSource
+from volur.sdk.v1alpha2.sources.csv.base import ProductsSource
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/client.py:7]: Both these lines could use the structure you are buidling, which is absolutely correct:

 ```suggestion
from volur.sdk.v1alpha1.sources.csv import MaterialsSource
from volur.sdk.v1alpha2.sources.csv import ProductsSource
```

File: src/volur/api/v1alpha1/client.py [hunk 7]
```diff
@@ -111,3 +112,93 @@ async def generate_requests() -> (
...
+                    yield product_pb2.UploadProductInformationRequest(
+                        product=product,
+                    )
+            except Exception:
+                logger.exception(
+                    "error occurred while generating requests",
+                )
+
+        try:
+            logger.info("start uploading materials data")
```
Comment (nikitabarskov) [inline src/volur/api/v1alpha1/client.py:149]:
```suggestion
            logger.info("start uploading products data")
```

File: src/volur/sdk/v1alpha2/__init__.py [hunk 8]
```diff
@@ -1,5 +1,11 @@
 from volur.sdk.v1alpha2.client import VolurClient
+from volur.sdk.v1alpha2.sources import (
+    ProductsCSVFileSource,
+    ProductsSource,
+)
```
Comment (nikitabarskov) [inline src/volur/sdk/v1alpha2/__init__.py:5]: Not really required, let sources live in `volur.sdk.v1alpha2.sources.csv` module for now

**Review by nikitabarskov — COMMENTED** [review]
Really good!

I left some comments, please address them.

**Optionally:** I would add an example for Products endpoint in Azure Functions and Local example `examples/main.py`, but it's not required.

PS: if you run test with `just test` or `earthly +test`, you will observe a folder `htmlcov`, if you open `htmlcov/index.html` you can see the test coverage, which in this specific scenario is 98% for ProductsCSVSource.

File: src/volur/sdk/v1alpha2/__init__.py (hunk 3 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/__init__.py]: Removed when merged with main

File: src/volur/api/v1alpha1/client.py (hunk 2 above)
Comment (rebwibsey) [reply src/volur/api/v1alpha1/client.py:128]: Updated

File: src/volur/sdk/v1alpha2/sources/__init__.py (hunk 4 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/sources/__init__.py]: Removed when merged with main

File: src/volur/sdk/v1alpha2/sources/csv/__init__.py (hunk 5 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/sources/csv/__init__.py:1]: Updated when merged with main

File: src/volur/sdk/v1alpha2/client.py (hunk 6 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/client.py]: Updated when merged with main

File: src/volur/api/v1alpha1/client.py (hunk 7 above)
Comment (rebwibsey) [reply src/volur/api/v1alpha1/client.py:149]: Updated

File: src/volur/sdk/v1alpha2/__init__.py (hunk 8 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/__init__.py]: Updated when merged with main

File: src/volur/sdk/v1alpha2/sources/csv/source.py (hunk 1 above)
Comment (rebwibsey) [reply src/volur/sdk/v1alpha2/sources/csv/source.py]: Yes, good point. Updated

**Review by nikitabarskov — APPROVED** [review]

**Pull request merged**